README.md
<div align="center">
  <img  height="200"
    src="./logo.jpg">
  <h1>jest-github-reporter</h1>
  <p>See jest test errors directly in pull requests</p>
</div>

## Highlights

- Report your jest results as annotations using the [GitHub Checks API](https://developer.github.com/v3/checks/)
- Only runs in CI environment

![Example of annotations being included in a pull request](./example.png)

## Install

```sh
npm install --save-dev jest-github-reporter
# or
yarn add -D jest-github-reporter
```

## Usage

You will need to install [the github app](https://github.com/apps/jest-results) to your repo.

Then just run jest and it will test errors PRs!

```json
{
  "reporters": ["default", "jest-github-reporter"]
}
```

```sh
jest --testLocationInResults file.js
```

(The `--testLocationInResults` flag is needed to get the in-file checks/annotations, otherwise only the summary comment is shown.)

## Using your own GitHub App

You might not want to use our github app for the formatter.

Reasons:

1. You think they way we exposed the keys is bad
2. You are on Github Enterprise and cannot use the public app

In these situations all you need to do is create a GitHub app and set a few environment variables.

### 1. Create a GitHub app

Go to [this page](https://github.com/settings/apps) to create a new GitHub app. You need to set the following fields:

- `name` - The name that shows for your app in the checks reports
- `Homepage URL` - You can set this to anything. Ours is set to `https://github.com`
- `Webhook URL` - You can set this to anything. Ours is set to `https://github.com`

Then hit `Save Changes` and you're all done setting up your GitHub app.

### 2. Set `JEST_APP_ID` environment variable

Your GitHub application's ID. This can be found at the top of your GitHub app's edit page.

### 3. Set `JEST_PRIVATE_KEY` environment variable

The private RSA key for your application. The prompt to generate the RSA key is at the bottom of your GitHub app's edit page.

Once you have generated a key, open the file that is downloaded and copy to text into the `PRIVATE_KEY` environment variable.

### 4. Set `GH_API` (enterprise only)

To get this package to work on github enterprise instances you will need to set the `GH_API` environment variable to a url pointing towards your enterprise GitHub's API.

If you need specific preview apis for your enterprise server, you can set the `GH_PREVIEWS` environment variable.

Example:

```
GH_PREVIEWS=eye-scream-preview,machine-man-preview
```

### 5. (optional) Set `GH_CHECK_NAME`

If the default check name conflicts with something, you can override it by passing `GH_CHECK_NAME` environment variable.


CHANGELOG.md
# v1.1.1 (Thu Mar 02 2023)

#### 🐛 Bug Fix

- annotate files relative to git root, instead of cwd, so that it works in nested packages in monorepos [#151](https://github.com/hipstersmoothie/jest-github-reporter/pull/151) ([@PaulGiletich](https://github.com/PaulGiletich))

#### Authors: 1

- [@PaulGiletich](https://github.com/PaulGiletich)

---

# v1.1.0 (Thu Mar 02 2023)

#### 🚀 Enhancement

- Allow to pass check name environment variable [#150](https://github.com/hipstersmoothie/jest-github-reporter/pull/150) ([@PaulGiletich](https://github.com/PaulGiletich))

#### Authors: 1

- [@PaulGiletich](https://github.com/PaulGiletich)

---

# v1.0.2 (Mon Feb 15 2021)

:tada: This release contains work from new contributors! :tada:

Thanks for all your work!

:heart: Theryn Groetken ([@therynamo](https://github.com/therynamo))

:heart: Øyvind Smestad ([@osmestad](https://github.com/osmestad))

#### 🐛 Bug Fix

- chore: Update Create Check [#142](https://github.com/hipstersmoothie/jest-github-reporter/pull/142) ([@therynamo](https://github.com/therynamo))

#### ⚠️ Pushed to `master`

- upgrade typescript ([@hipstersmoothie](https://github.com/hipstersmoothie))
- upgrade auto ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### 📝 Documentation

- Improve docs on how to get the test result details reported [#138](https://github.com/hipstersmoothie/jest-github-reporter/pull/138) ([@osmestad](https://github.com/osmestad))

#### Authors: 3

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Øyvind Smestad ([@osmestad](https://github.com/osmestad))
- Theryn Groetken ([@therynamo](https://github.com/therynamo))

---

# v1.0.1 (Wed May 27 2020)

#### 🐛 Bug Fix

- Update README.md correct usage instructions [#117](https://github.com/hipstersmoothie/jest-github-reporter/pull/117) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v1.0.0 (Mon May 11 2020)

#### 💥 Breaking Change

- Force major: drop node 8 [#106](https://github.com/hipstersmoothie/jest-github-reporter/pull/106) ([@hipstersmoothie](https://github.com/hipstersmoothie))

#### Authors: 1

- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))

---

# v0.4.7 (Mon May 11 2020)

:tada: This release contains work from a new contributor! :tada:

Thank you, WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot)), for all your work!

#### 🐛 Bug Fix

- upgrade create-check [#105](https://github.com/hipstersmoothie/jest-github-reporter/pull/105) ([@hipstersmoothie](https://github.com/hipstersmoothie))
- Update dependency jest to v26 [#104](https://github.com/hipstersmoothie/jest-github-reporter/pull/104) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency eslint-plugin-jsdoc to v25 [#102](https://github.com/hipstersmoothie/jest-github-reporter/pull/102) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))

#### Authors: 3

- [@renovate[bot]](https://github.com/renovate[bot])
- Andrew Lisowski ([@hipstersmoothie](https://github.com/hipstersmoothie))
- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.4.6 (Mon May 11 2020)

:tada: This release contains work from a new contributor! :tada:

Thank you, WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot)), for all your work!

#### 🐛 Bug Fix

- Update typescript-eslint monorepo to v2.32.0 [#94](https://github.com/hipstersmoothie/jest-github-reporter/pull/94) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency husky to v4.2.5 [#97](https://github.com/hipstersmoothie/jest-github-reporter/pull/97) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency eslint-plugin-import to v2.20.2 [#96](https://github.com/hipstersmoothie/jest-github-reporter/pull/96) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency lint-staged to v10 [#89](https://github.com/hipstersmoothie/jest-github-reporter/pull/89) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))

#### Authors: 2

- [@renovate[bot]](https://github.com/renovate[bot])
- WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot))

---

# v0.4.5 (Mon May 11 2020)

:tada: This release contains work from a new contributor! :tada:

Thank you, WhiteSource Renovate ([@renovate-bot](https://github.com/renovate-bot)), for all your work!

#### 🐛 Bug Fix

- Update dependency eslint to v7 [#101](https://github.com/hipstersmoothie/jest-github-reporter/pull/101) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency eslint-plugin-prettier to v3.1.3 [#99](https://github.com/hipstersmoothie/jest-github-reporter/pull/99) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency @types/node to v12.12.38 [#95](https://github.com/hipstersmoothie/jest-github-reporter/pull/95) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency eslint-config-prettier to v6.11.0 [#93](https://github.com/hipstersmoothie/jest-github-reporter/pull/93) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))
- Update dependency auto to v9.31.2 [#91](https://github.com/hipstersmoothie/jest-github-reporter/pull/91) ([@renovate-bot](https://github.com/renovate-bot) [@renovate[bot]](https://github.com/renovate[bot]))

example/test.js
test('test', () => {
  expect(3).toBe(2);
});

example/another.test.js
test('test', () => {
  expect(1).toBe(2);
});


src/groupSequences.test.ts
import groupSequences from './groupSequences';

describe('groupSequences', () => {
  it('should return an empty array when array is empty', () => {
    expect(groupSequences([])).toEqual([]);
  });
  it('should return grouped sequences', () => {
    const array = [1, 2, 3, 10, 11, 12, 13, 42];

    expect(groupSequences(array)).toEqual([
      [1, 2, 3],
      [10, 11, 12, 13],
      [42]
    ]);
  });
})

src/auth.ts
import { request } from '@octokit/request';
import { App } from '@octokit/app';
import Octokit from '@octokit/rest';
import envCi from 'env-ci';

const APP_ID = process.env.JEST_APP_ID
  ? Number(process.env.JEST_APP_ID)
  : 38833;
/**
 * Before you say anything I *know* this is horribly insecure.
 *
 * If we were not to to this then every user would have to create
 * their own GitHub App and manage the APP_ID and PRIVATE_KEY through
 * env vars.
 *
 * How could this go wrong? Well this PRIVATE_KEY only creates jwt
 * tokens that work on people who have installed the Jest Results
 * App. If an attacker got ahold of the token they could only read repo
 * metadata and read/write checks. So the attack surface is really only
 * messing with a users checks, which is not too risky.
 */
const PRIVATE_KEY =
  process.env.JEST_PRIVATE_KEY ||
  `-----BEGIN RSA PRIVATE KEY-----
MIIEogIBAAKCAQEAwKPGnXvRdbkXXaIzw2YV4ykKCx6Lx1N+FAByUvQ5k7XaaSsi
X7MFx3XJlUVMNf47ur5NQ2KjkDqA4Q9alb99yxUzSnsLBsHWQKEUZKmP86dqe2Ku
a2XG+GYtUnaaRrAsj8BWqFZ1Hq38P8ge8BIA51+pgpYOsX3imT5PhJfloeMXCMQ4
GU+nJf8s0DzxRav3THzDKfzbQ487B1CbcTcEM70v+lFCKX36AzssYnKiEEyDOlUO
6HkwQB+CTvQgB51JatmFFFZfCgnjYPQHtO6YqFWn8QBF0s9wHhoHWBNx+ZKoOVKw
qOKcFwYmHdAzpuLmt0gGp4ewBswImJVpsekBQQIDAQABAoIBADp5/LKVgXHQ68za
jggElyRLsubJMPki2STNNecEh+3UyCYgl+ChAWIY2UZcsNO7BvqbBe0spiYD/FdV
R9QpOtBI8Tbsvt4gPR+FRiGAb1gxO9uUiwnC7XE94wgjRJWsqPpCEowrIoZbnjTm
VK3faTLTESu4zWEHq5+FELJZQbWwJ/SjJkA1viizBLSLCnhlMzsabHZHKtLPvZJ0
tnRPVidp1hzQzDOIQBdojUvMZ3evLqYL1pwoRnqitotuchDprTUqGfBLFy1A+1iK
QqPpHnDe8oEeITOi8cJhOmyk2ZmkSIyRmqZi9Cf/lItQBKws8aHUIoex7AfpXK30
1mP1wAECgYEA+WIhAzEj7wXu/J+fPfZs0eW8htGtxQTQYpLztfxKGqS4spGJcj89
ga0MNvGAexNMsWW+n9SHNhVuJB7XSuxMrvXPuMPfwT1u910Ah/Wh+61saAOHRA+O
3c+OyI2ojTchsoC3k1VZ0RTeDZXL870+wLmf1MqEOGovvnUx10oGfcECgYEAxcA7
JJko/IbkCECtxcupIieZJdELDSkRdFZgVVHZcxwujJxyj2O/YZB31c6AaCbWnIO3
s98SPhLLS4b10pEqR8FbCKKm9MiqWybAeWn14IM4EAgGRReWquTOWv/bDLjKphTr
oMm4lBm1dXrOfIuNyTOAJPIjJvtUCjfAroaCY4ECgYBllKIL0c1oREt3nXFY5PKo
gOLNK8WTdgWH0YHyBAUPWz9chUmuPrJICvvpuW9zMoZP0DjYk9JLpmkJz4I0o5IM
xlXJVgfjh6mWmsxnlRdZE+gPajiD8a5pDW2EpacddnKEakfcfKysLMrST80WyGQy
TqobHC7FaANwmf4mSqHgQQKBgBwgCAlfgzXPVZVa5ZwxKCAEc8KuJZ08jw/1zQO1
fXDivDghdCWysSCGNJUDJr4pb/KYxULe4jBT6fgW/NVy8gl8lZ73yzkbZSdLrqpW
CLNi1lFpYsLm5PXvTu4gX55ClgfjB9Q5fHgL5AQOcFnEW2kXWw1mJtu/eSdu4Iex
98+BAoGAOF4GUl/Aa6rJNgjBUNZjUomV+FJL3htuu1b3t2tLGnozPxFS6RPMWLvd
AMBi+Wv6FzuOhZx6aIMlz/LEgC+s9o3S+mSlsp0/UoNcy7j1zSosb4C8QqXbTGRG
sUdulK+rOI5aef3CRA2/j6V1expVZF/ttOsmybZcuUcSGxN1N5Q=
-----END RSA PRIVATE KEY-----
`;

const env = envCi();
const app = new App({ id: APP_ID, privateKey: PRIVATE_KEY });
const jwt = app.getSignedJsonWebToken();
const [owner = 'hipstersmoothie', repo = 'jest-github-reporter'] =
  'slug' in env ? env.slug.split('/') : [];

export default async function authenticateApp() {
  const { data } = await request('GET /repos/:owner/:repo/installation', {
    owner,
    repo,
    headers: {
      authorization: `Bearer ${jwt}`,
      accept: 'application/vnd.github.machine-man-preview+json'
    }
  });

  const installationId = data.id;
  const token = await app.getInstallationAccessToken({
    installationId
  });

  return new Octokit({
    auth: token,
    previews: ['symmetra-preview']
  });
}


src/StdIo.ts
export default class StdIo {
  private _out: typeof process.stdout.write;
  private _err: typeof process.stderr.write;
  private _bufferedOutput: Set<() => void>;

  constructor() {
    this._out = process.stdout.write.bind(process.stdout);
    this._err = process.stderr.write.bind(process.stderr);
    this._bufferedOutput = new Set<() => void>();

    this._wrapStdio(process.stdout);
    this._wrapStdio(process.stderr);
  }

  log(message: string) {
    process.stderr.write(message + '\n');
  }

  close() {
    this._forceFlushBufferedOutput();
    process.stdout.write = this._out;
    process.stderr.write = this._err;
  }

  // Don't wait for the debounced call and flush all output immediately.
  _forceFlushBufferedOutput() {
    for (const flushBufferedOutput of this._bufferedOutput) {
      flushBufferedOutput();
    }
  }

  _wrapStdio(stream: NodeJS.WriteStream) {
    const originalWrite = stream.write;

    let buffer: string[] = [];
    let timeout: NodeJS.Timeout | null = null;

    const flushBufferedOutput = () => {
      const string = buffer.join('');
      buffer = [];

      if (string) {
        originalWrite.call(stream, string);
      }

      this._bufferedOutput.delete(flushBufferedOutput);
    };

    this._bufferedOutput.add(flushBufferedOutput);

    const debouncedFlush = () => {
      // If the process blows up no errors would be printed.
      // There should be a smart way to buffer stderr, but for now
      // we just won't buffer it.
      if (stream === process.stderr) {
        flushBufferedOutput();
      } else {
        if (!timeout) {
          timeout = setTimeout(() => {
            flushBufferedOutput();
            timeout = null;
          }, 100);
        }
      }
    };

    stream.write = (chunk: string) => {
      buffer.push(chunk);
      debouncedFlush();
      return true;
    };
  }
}


src/create-check.ts
// Only Types
// eslint-disable-next-line import/no-extraneous-dependencies
import { TestResult, AggregatedResult } from '@jest/reporters';

import path from 'path';
import stripAnsi from 'strip-ansi';
import createCheck, { Annotation } from 'create-check';
import execa from 'execa';

const APP_ID = 38833;
/**
 * Before you say anything I *know* this is horribly insecure.
 *
 * If we were not to to this then every user would have to create
 * their own GitHub App and manage the APP_ID and PRIVATE_KEY through
 * env vars.
 *
 * How could this go wrong? Well this PRIVATE_KEY only creates jwt
 * tokens that work on people who have installed the Jest Results
 * App. If an attacker got ahold of the token they could only read repo
 * metadata and read/write checks. So the attack surface is really only
 * messing with a users checks, which is not too risky.
 */
const PRIVATE_KEY = `-----BEGIN RSA PRIVATE KEY-----
MIIEogIBAAKCAQEAwKPGnXvRdbkXXaIzw2YV4ykKCx6Lx1N+FAByUvQ5k7XaaSsi
X7MFx3XJlUVMNf47ur5NQ2KjkDqA4Q9alb99yxUzSnsLBsHWQKEUZKmP86dqe2Ku
a2XG+GYtUnaaRrAsj8BWqFZ1Hq38P8ge8BIA51+pgpYOsX3imT5PhJfloeMXCMQ4
GU+nJf8s0DzxRav3THzDKfzbQ487B1CbcTcEM70v+lFCKX36AzssYnKiEEyDOlUO
6HkwQB+CTvQgB51JatmFFFZfCgnjYPQHtO6YqFWn8QBF0s9wHhoHWBNx+ZKoOVKw
qOKcFwYmHdAzpuLmt0gGp4ewBswImJVpsekBQQIDAQABAoIBADp5/LKVgXHQ68za
jggElyRLsubJMPki2STNNecEh+3UyCYgl+ChAWIY2UZcsNO7BvqbBe0spiYD/FdV
R9QpOtBI8Tbsvt4gPR+FRiGAb1gxO9uUiwnC7XE94wgjRJWsqPpCEowrIoZbnjTm
VK3faTLTESu4zWEHq5+FELJZQbWwJ/SjJkA1viizBLSLCnhlMzsabHZHKtLPvZJ0
tnRPVidp1hzQzDOIQBdojUvMZ3evLqYL1pwoRnqitotuchDprTUqGfBLFy1A+1iK
QqPpHnDe8oEeITOi8cJhOmyk2ZmkSIyRmqZi9Cf/lItQBKws8aHUIoex7AfpXK30
1mP1wAECgYEA+WIhAzEj7wXu/J+fPfZs0eW8htGtxQTQYpLztfxKGqS4spGJcj89
ga0MNvGAexNMsWW+n9SHNhVuJB7XSuxMrvXPuMPfwT1u910Ah/Wh+61saAOHRA+O
3c+OyI2ojTchsoC3k1VZ0RTeDZXL870+wLmf1MqEOGovvnUx10oGfcECgYEAxcA7
JJko/IbkCECtxcupIieZJdELDSkRdFZgVVHZcxwujJxyj2O/YZB31c6AaCbWnIO3
s98SPhLLS4b10pEqR8FbCKKm9MiqWybAeWn14IM4EAgGRReWquTOWv/bDLjKphTr
oMm4lBm1dXrOfIuNyTOAJPIjJvtUCjfAroaCY4ECgYBllKIL0c1oREt3nXFY5PKo
gOLNK8WTdgWH0YHyBAUPWz9chUmuPrJICvvpuW9zMoZP0DjYk9JLpmkJz4I0o5IM
xlXJVgfjh6mWmsxnlRdZE+gPajiD8a5pDW2EpacddnKEakfcfKysLMrST80WyGQy
TqobHC7FaANwmf4mSqHgQQKBgBwgCAlfgzXPVZVa5ZwxKCAEc8KuJZ08jw/1zQO1
fXDivDghdCWysSCGNJUDJr4pb/KYxULe4jBT6fgW/NVy8gl8lZ73yzkbZSdLrqpW
CLNi1lFpYsLm5PXvTu4gX55ClgfjB9Q5fHgL5AQOcFnEW2kXWw1mJtu/eSdu4Iex
98+BAoGAOF4GUl/Aa6rJNgjBUNZjUomV+FJL3htuu1b3t2tLGnozPxFS6RPMWLvd
AMBi+Wv6FzuOhZx6aIMlz/LEgC+s9o3S+mSlsp0/UoNcy7j1zSosb4C8QqXbTGRG
sUdulK+rOI5aef3CRA2/j6V1expVZF/ttOsmybZcuUcSGxN1N5Q=
-----END RSA PRIVATE KEY-----
`;

interface Location {
  column: number;
  line: number;
}

async function createAnnotations(results: TestResult[]) {
  const repoRoot = (await execa('git', ['rev-parse', '--show-toplevel'])).stdout;

  const annotations: Annotation[] = [];

  for (const result of results) {
    const { testFilePath, testResults } = result;

    for (const failure of testResults) {
      if ('location' in failure) {
        const { location = {} as Location, failureMessages } = failure as {
          failureMessages: string[];
          location: Location;
        };

        if (location) {
          failureMessages.forEach(message => {
            const numbers = message.match(
              new RegExp(`${result.testFilePath}:(\\d+):\\d+`)
            );
            const start_line = numbers
              ? Number(numbers[1])
              : location.line || 0;

            annotations.push({
              path: path.relative(repoRoot, testFilePath),
              start_line,
              end_line: start_line,
              annotation_level: 'failure',
              message: failureMessages.map(stripAnsi).join('\n')
            });
          });
        }
      }
    }
  }

  return annotations;
}

export default async (results: AggregatedResult) =>
  createCheck({
    tool: 'Jest',
    name: process.env.GH_CHECK_NAME || 'Test',
    annotations: await createAnnotations(results.testResults),
    errorCount: results.numFailedTests,
    appId: process.env.JEST_APP_ID ? Number(process.env.JEST_APP_ID) : APP_ID,
    privateKey: process.env.JEST_PRIVATE_KEY || PRIVATE_KEY
  });
